### PR TITLE
Feature/task/419381 create generic frontend controller

### PR DESCRIPTION
**Pull request by zozanKirkutay**

### CODE REVIEW

File: src/EPR.Payment.Portal/Services/PaymentsService.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        public PaymentsService(IMapper mapper, IHttpPaymentFacade httpPaymentFacade)
+        {
+            _mapper = mapper;
+            _httpPaymentFacade = httpPaymentFacade ?? throw new ArgumentNullException(nameof(httpPaymentFacade));
+        }
+        public async Task<CompletePaymentViewModel> CompletePaymentAsync(Guid externalPaymentId, CancellationToken cancellationToken)
+        {
+            if (externalPaymentId == Guid.Empty)
+            {
+                throw new ArgumentException(ExceptionMessages.ErrorEcternalPaymentIdEmpty);
```
Comment (paulshimmen) [inline src/EPR.Payment.Portal/Services/PaymentsService.cs:23]: Small spelling mistake here. Please change to:

throw new ArgumentException(ExceptionMessages.ErrorExternalPaymentIdEmpty);

File: src/EPR.Payment.Portal/Services/PaymentsService.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+                var completePaymentResponseDto = await _httpPaymentFacade.CompletePaymentAsync(externalPaymentId, cancellationToken);
+                if (string.IsNullOrEmpty(completePaymentResponseDto.Reference))
+                {
+                    throw new Exception(ExceptionMessages.PaymentDataNotFound);
+                }
+                var viewModel = _mapper.Map<CompletePaymentViewModel>(completePaymentResponseDto);
+                return viewModel;
+            }
+            catch (Exception ex)
+            {
```
Comment (paulshimmen) [inline src/EPR.Payment.Portal/Services/PaymentsService.cs:36]: I think we should add logging here.

File: src/EPR.Payment.Portal/Views/PaymentSuccess/Index.cshtml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿@using EPR.Payment.Portal.Common.Configuration
+@using Microsoft.AspNetCore.Mvc.Routing;
+@model DashboardConfiguration;
```
Comment (paulshimmen) [inline src/EPR.Payment.Portal/Views/PaymentSuccess/Index.cshtml:3]: I dont see this mode being used here. If it is not needed, please remove it.

File: src/EPR.Payment.Portal/Views/PaymentSuccess/Index.cshtml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+
+@{
+    ViewData["Title"] = "GovPay Success";
+    Layout = "~/Views/Shared/_Layout.cshtml";
+}
+
+<main class="govuk-main-wrapper govuk-!-padding-top-4" id="main-content">
```
Comment (paulshimmen) [inline src/EPR.Payment.Portal/Views/PaymentSuccess/Index.cshtml:11]: As a suggestion, we could add role="main" here which will help with accessibility.

File: src/EPR.Payment.Portal.Common/Constants/ExceptionMessages.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    public static class ExceptionMessages
+    {
+        // HttpPaymentsFacade exceptions
+        public const string PaymentFacadeBaseUrlMissing = "PaymentFacade BaseUrl configuration is missing";
+        public const string PaymentFacadeEndPointNameMissing = "PaymentFacade EndPointName configuration is missing";
+        public const string PaymentFacadeHttpClientNameMissing = "PaymentFacade HttpClientName configuration is missing";
+        public const string ErrorRetrievingCompletePayment = "Error occured while retrieving complete payment.";
+        public const string ErrorCompletePayment = "Error occured while completing payment.";
+        public const string PaymentDataNotFound = "Payment data not found.";
+        public const string ErrorEcternalPaymentIdEmpty = "Payment data not found.";
```
Comment (paulshimmen) [inline src/EPR.Payment.Portal.Common/Constants/ExceptionMessages.cs:18]: Please change to ErrorExternalPaymentIdEmpty

File: src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    public class GovPayCallbackController : Controller
+    {
+        private readonly IPaymentsService _paymentsServices;
+
+        public GovPayCallbackController(IPaymentsService paymentsServices)
+        {
+            _paymentsServices = paymentsServices;
+        }
+        public async Task<IActionResult> Index(Guid id, CancellationToken cancellationToken)
+        {
```
Comment (paulshimmen) [inline src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs:17]: I think we should add a check for the id here. ie: 

 if (id == Guid.Empty)
            {
                // Return an error view if the ID is invalid
                return RedirectToAction("Index", "PaymentError", new { message = "Invalid payment ID." });
            }

File: src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            try
+            {
+                var viewModel = await _paymentsServices.CompletePaymentAsync(id, cancellationToken);
+                string controllerName = string.Concat("Payment", ((PaymentStatus)viewModel.Status).ToString());
+                return RedirectToAction("Index", controllerName, viewModel);
+            }
+            catch (Exception ex)
+            {
```
Comment (paulshimmen) [inline src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs:25]: Add logging here please.

File: src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+using Microsoft.AspNetCore.Mvc;
+
+namespace EPR.Payment.Portal.Controllers
+{
+    public class GovPayCallbackController : Controller
+    {
+        private readonly IPaymentsService _paymentsServices;
+
+        public GovPayCallbackController(IPaymentsService paymentsServices)
+        {
```
[the remark this answers is not in the document]
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs:14]: logger was added

File: src/EPR.Payment.Portal/Controllers/PaymentSuccessController.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        private readonly DashboardConfiguration _dashboardConfiguration;
+
+        public PaymentSuccessController(IOptions<DashboardConfiguration> dashboardConfiguration)
+        {
+            _dashboardConfiguration = dashboardConfiguration.Value ?? throw new ArgumentNullException(nameof(dashboardConfiguration));
+        }
+
+        public IActionResult Index(CompletePaymentViewModel completePaymentResponseViewModel)
+        {
```
[the remark this answers is not in the document]
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Controllers/PaymentSuccessController.cs:10]: added null control for completePaymentResponseViewModel

File: src/EPR.Payment.Portal/Services/PaymentsService.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+namespace EPR.Payment.Portal.Services
+{
+    public class PaymentsService : IPaymentsService
+    {
+        private readonly IHttpPaymentFacade _httpPaymentFacade;
+        private IMapper _mapper;
+
+        public PaymentsService(IMapper mapper, IHttpPaymentFacade httpPaymentFacade)
+        {
+            _mapper = mapper;
```
[the remark this answers is not in the document]
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Services/PaymentsService.cs]: added _mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));

File: src/EPR.Payment.Portal/Services/PaymentsService.cs (hunk 1 above)
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Services/PaymentsService.cs]: Fixed spelling mistake

File: src/EPR.Payment.Portal/Views/PaymentSuccess/Index.cshtml (hunk 3 above)
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Views/PaymentSuccess/Index.cshtml]: deleted @model DashboardConfiguration

File: src/EPR.Payment.Portal/Views/PaymentSuccess/Index.cshtml (hunk 4 above)
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Views/PaymentSuccess/Index.cshtml:9]: we agreed not to add it

File: src/EPR.Payment.Portal.Common/Constants/ExceptionMessages.cs (hunk 5 above)
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal.Common/Constants/ExceptionMessages.cs]: corrected ErrorEcternalPaymentIdEmpty

File: src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs (hunk 6 above)
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs:19]: added id check

File: src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs (hunk 7 above)
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs:33]: added logging to catch

File: src/EPR.Payment.Portal/Services/PaymentsService.cs (hunk 2 above)
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Services/PaymentsService.cs:39]: added logging

File: src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+            _paymentsServices = paymentsServices;
```
[the remark this answers is not in the document]
Comment (zozanKirkutay) [reply src/EPR.Payment.Portal/Controllers/GovPayCallbackController.cs]: added _paymentsService = paymentsService ?? throw new ArgumentNullException(nameof(paymentsService));

**Review by paulshimmen — APPROVED** [review]

**Pull request merged**